spinn_machine/base_multicast_routing_entry.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2014 The University of Manchester
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+from typing import Iterable, Optional, Set, Union
+from spinn_machine.constants import MAX_LINKS_PER_ROUTER
+from spinn_machine.data import MachineDataView
+
+
+class BaseMulticastRoutingEntry(object):
+    """
+    Represents an entry in a SpiNNaker chip's multicast routing table.
+    There can be up to 1024 such entries per chip, but some may be reserved
+    for system purposes.
+    """
+
+    # cache of MachineDataView.get_machine_version().max_cores_per_chip
+    max_cores_per_chip: int = 0
+
+    __slots__ = ["_spinnaker_route"]
+
+    def __init__(self, processor_ids:  Union[int, Iterable[int], None],
+                 link_ids:  Union[int, Iterable[int], None],
```

Comment:
Can't quite remember if we prefer Optional rather than Union[..., None], though if the comment below is taken into account, this would of course go away...